streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/RackAwareTaskAssignor.java

```diff
@@ -185,4 +193,212 @@ public boolean validateClientRack() {
+        }
+
+        final int sinkId = getSinkID(clientList, taskIdList);
+        for (int taskNodeId = 0; taskNodeId < taskIdList.size(); taskNodeId++) {
+            graph.addEdge(SOURCE_ID, taskNodeId, 1, 0, 1);
+        }
+
+        // It's possible that some clients have 0 task assign. These clients will have 0 tasks assigned
+        // even though it may have lower traffic cost. This is to maintain the balance requirement.
+        for (int i = 0; i < clientList.size(); i++) {
+            final int clientNodeId = taskIdList.size() + i;
+            final int capacity = originalAssignedTaskNumber.getOrDefault(clientList.get(i), 0);
+            // Flow equals to capacity for edges to sink
+            graph.addEdge(clientNodeId, sinkId, capacity, 0, capacity);
+        }
+
+        graph.setSourceNode(SOURCE_ID);
+        graph.setSinkNode(sinkId);
+
+        return graph;
+    }
+
+    private void assignActiveTaskFromMinCostFlow(final Graph<Integer> graph,
+                                                 final SortedSet<TaskId> activeTasks,
+                                                 final List<UUID> clientList,
+                                                 final List<TaskId> taskIdList,
+                                                 final Map<UUID, ClientState> clientStates,
+                                                 final Map<UUID, Integer> originalClientCapacity,
+                                                 final Map<TaskId, UUID> taskClientMap) {
+        int taskAssigned = 0;
+        for (int taskNodeId = 0; taskNodeId < taskIdList.size(); taskNodeId++) {
+            final TaskId taskId = taskIdList.get(taskNodeId);
+            final Map<Integer, Graph<Integer>.Edge> edges = graph.edges(taskNodeId);
+            for (final Graph<Integer>.Edge edge : edges.values()) {
+                if (edge.flow > 0) {
+                    taskAssigned++;
+                    final int clientIndex = edge.destination - taskIdList.size();
+                    final UUID clientId = clientList.get(clientIndex);
+                    final UUID originalClientId = taskClientMap.get(taskId);
+
+                    // Don't need to assign this task to other client
+                    if (clientId.equals(originalClientId)) {
+                        break;
+                    }
+
+                    final ClientState clientState = clientStates.get(clientId);
+                    final ClientState originalClientState = clientStates.get(originalClientId);
+                    originalClientState.unassignActive(taskId);
+                    clientState.assignActive(taskId);
+                }
+            }
+        }
+
+        // Validate task assigned
+        if (taskAssigned != activeTasks.size()) {
+            throw new IllegalStateException("Computed active task assignment number "
+                + taskAssigned + " is different size " + activeTasks.size());
+        }
+
+        // Validate capacity constraint
```

Comment:
Is this really a "capacity" or rather the "load" of a client?

Reply:
This is the `originalAssignedTaskNum`. "load" is more accurate than `capacity`